PowerShell Core Policy (#180)
PowerShell/PowerShell-RFC: 1-Draft/RFCXXXX-Policy.md

---
RFC:  RFCnnnn
Author:  travisez13
Status:  Draft
SupercededBy:  N/A
Version: 0.1
Area: Engine
Comments Due: 6/30/2019
---

# `PowerShell Core` Policy

## Motivation

> Consumers, developers, and enterprise system administrators should be able to flexibly and reliable configure PowerShell 7.

Reviewer: "reliably"

Author: ```suggestion
Consumers, developers, and enterprise system administrators should be able to flexibly and reliably configure PowerShell 7.
```

Author: Please start using suggestion for this type of thing:
![image](https://user-images.githubusercontent.com/10873629/58736044-49ac7600-83b1-11e9-8b74-65e953c4360e.png)

Reviewer: I apologize. I'm not good with GitHub, yet.

## Acknowledgement

I based this off of @iSazonov 's RFC, for just a slightly different purpose.

> [PR #111](https://github.com/PowerShell/PowerShell-RFC/pull/111)

Author: Your RFC proposed adding an API.  This is only the subset of fixing the way policy works.  

I think we need to separate the two.  If you close, yours and re-open it after this one, that would work.

> [PR #111](https://github.com/PowerShell/PowerShell-RFC/pull/111)

Author: @iSazonov You haven't objected.  So, I'm going to resolve this.

## Goals

1. Specify how PowerShell 7 will deal with having both Windows PowerShell and PowerShell Core Group Policy.
    - This is covered in [Policy settings Setting Fall-Back](#policy-settings-setting-fall-back).

> 1. Correct how the `pwsh -settingsfile` switch behaves.

Reviewer: ```suggestion
1. Define how the `pwsh -settingsfile` switch behaves.
```

Author: Since it already behaves one way, that is not quite correct.  
```suggestion
1. Define how the `pwsh -settingsfile` switch should behave.
```

    - This is covered in  [Parameter `-settingsfile`](#Parameter--settingsfile)

## Definitions

> - **Computer-Wide settings/policy**  - setting or policy applied to the entire operating system.

Reviewer: applied to an operating system environment (OSE), including all users of the OSE

Author: ```suggestion
- **Computer-Wide settings/policy**  - setting or policy applied to an operating system environment (OSE), affecting all users of the OSE.
```

> - **User settings/policy** - setting or policy applied only to the user.

Reviewer: applied to a specific user of an OSE, not applied to the OSE as a whole

Author: it would be nice if you used the suggestions feature.
```suggestion
- **User settings/policy** - setting or policy applied to a specific user of an OSE, and not applied to the OSE as a whole.
```

I think this language is very technical.

## Specification

`PowerShell 7` should be configured using the following schemes:

> - On Windows - Group Policy Objects (GPO), Group Policy Preferences (GPP) and settings files.

Reviewer: does this imply an order hierarchy?

Author: precedence is described in the precedence section.  This is just introducing the concepts.

Author: ```suggestion
- On Windows - Group Policy Objects (GPO), and settings files.
```
Removing the mention of GPP as it's basically just a way to set registry settings. 
I don't think the precedence has to be mentioned at every point we talk about these two things.

- On Unix - settings files.

The settings files have `JSON` format.

**Warning** The settings files differ from `PowerShell 7` _profile_ files, which are PowerShell scripts run at startup.

Configuration schemes allow to customize `PowerShell 7` in the most flexible way:

- Enterprise system administrators can use GPO,
    GPP and computer-wide settings files to apply approved configuration settings and mandatory security settings in a centralized manner.

>     Most settings can be applied either to the user or computer-wide.

Reviewer: if there is a specific hierarchy, it would be good to document it as such.

Author: I think this is technically covered in Registry keys and settings.  
```suggestion
    Most settings can be applied either to the user or computer-wide.  We will cover the precedence of these in [Registry keys and settings](#registry-keys-and-settings)
```

- Developers and consumers can use user, or computer-wide level setting files.

### Configuration defaults

> PowerShell 7 has hard-coded defaults for all configuration options.

Reviewer: options or settings?

Author: ```suggestion
PowerShell 7 has hard-coded defaults for all configuration policies and settings.
```

> ### Settings locations

Reviewer 2: We could have different files for policies and regular settings (powershell.config.policy.json and powershell.config.json). This allow us to protect user policy config file by setting file security permissions.

Author: Both of these files should be protected by permissions.   Separating the files, only adds complexity

Reviewer 2: %APPDATA%/powershell.config.json too?

Reviewer 2: @TravisEz13 Can you address my previous comment https://github.com/PowerShell/PowerShell-RFC/pull/180/files#r288379736 ?

The default values must be `secure-by-default`.

For release versions hard-coded defaults must be the same as ones in pre-installed configuration files. For preview versions they may vary (ex., enable experimental features and so on).

Computer-wide configuration includes security sensitive setting,

> and failing to read those setting could result in an insecure system.

Reviewer: perhaps "apply those settings" instead of "read those setting".

Author: No, this is a mis-reading of this section.

Author: ```suggestion
and failing to read those setting, for example if the file is locked or corrupted, could result in an insecure system.
```

So, if during startup, PowerShell 7 cannot read files read from the Computer-Wide scope,

> it fails to startup.

Reviewer: If during the startup process PowerShell 7 cannot read settings file from the Computer-Wide scope, PowerShell 7 will fail to initialize.

What about group policies?

Author: This should not happen.  If it does, we will probably not know because the error will be in group policy.

Author: @swngdnz can you acknowledge this?   Do you think this should be mentioned in the RFC?

Reviewer: I think you should mention that "Computer-Wide scope" also applies to group policies, yes.

Otherwise, it's cool.

Author: I added another line with a comment about registry/GP read issues.

If during startup PowerShell 7 cannot read user configuration files it uses _hardcoded_ defaults.

> If during operation PowerShell 7 cannot read configuration files it continue to use _current_ (runtime) configuration values.

Reviewer: I don't think I understand this. If policy files change during a specific session, PS7 will update to those new files and change behavior?

Author: Not a current behavior.  If it does, this is saying that updating the configuration is not a blocking operation.  I can cut this from the spec if you think it would clarify.  cc @iSazonov

Author: We also might not end up implementing every new specified behavior.  This for example could not be implemented without a significant issue.

Reviewer: Yes, that's why I was confused. If you will not update behavior with all changes, then I don't think you should process them at all.

So instead of the current verbiage, perhaps "PowerShell 7 does not update configuration values from modified configuration files during a given session." That would certainly simplify life.

Author: ```suggestion
PowerShell 7 does not update configuration values from modified configuration files during a given session after the configuration has been loaded.
```

### Settings locations

`PowerShell 7` settings are grouped into `Policy settings` and `Regular settings`.
Regular settings are normal configuration settings.
Regular settings can be treated as default and recommended values.

> Policy settings is higher precedence.

Reviewer: Policy settings have a higher precedence than regular settings.

Author: ```suggestion
Policy settings have a higher precedence than regular settings.
```

Author: @swngdnz Please, Please start using suggestions
![image](https://user-images.githubusercontent.com/10873629/58736259-08689600-83b2-11e9-803c-d7280abb5cd1.png)

For Policy Settings,
each policy should have a `Use Windows PowerShell Policy` which will indicate that the policy should the read from

> `SOFTWARE\Policies\Microsoft\Windows\PowerShell` instead of `Software\Policies\PowerShellCore`.

Reviewer 2: What is default?

Author: default in GP is no have no policy, so this behavior would not occur.

### Precedence of applying settings

Because a configuration setting can be in several schemes, the setting wins according to the priority of its scheme.

#### Precedence for Policy settings in descending order

| Scheme                      | Windows                                              | Unix                                                 |
|-----------------------------|------------------------------------------------------|------------------------------------------------------|
| GPO -> Computer Policy      | HKLM\Software\Policies\PowerShellCore                | /etc/powershell.config.json                          |
| GPO -> User Policy          | HKCU\Software\Policies\PowerShellCore                | See `Comment A` below                                |

> | File -> Computer-Wide       | %ProgramFiles%/PowerShell/powershell.config.json     | /opt/Microsoft/powershell/powershell.config.json     |

Reviewer 2: Why do we use "Microsoft" in the path?

Author: This should be in a `/etc` path.  I think this and the one on line 88 should be in `/etc/powershell/powershell.config.json`

Author: ```suggestion
| File -> Computer-Wide       | %ProgramFiles%/PowerShell/powershell.config.json     | /etc/powershell/powershell.config.json     |
```

| File -> Application-Startup | pwsh -settingsfile `somepath/powershell.config.json` | pwsh -settingsfile `somepath/powershell.config.json` |
| File -> User-Wide           | %APPDATA%/powershell.config.json                     | %XDG_CONFIG_HOME%/powershell.config.json             |
| File -> Application-Wide    | $apphome/powershell.config.json                      | $apphome/powershell.config.json                      |

Defaults:

`%APPDATA%` - `C:\Users\useraccount\AppData\Roaming`

`%XDG_CONFIG_HOME%` - `HOME/.config`

> 

Reviewer 2: I think we need add comment?example here about `$apphome` possible value(s).

Author: Please clarify?

Reviewer 2: Above we use `$apphome`. Here we could add short description of the term and example.

Author: ```suggestion
`$PSHome` definition from [about_powershell_config](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_powershell_config?view=powershell-6) - The $PSHOME location is defined as the same directory as the executing System.Management.Automation.dll assembly.
This applies to hosted PowerShell SDK instances as well.
```

#### Parameter `-settingsfile`

With `-settingsfile` parameter users can assign custom settings from the config file and overwrite user-wide and application-wide settings.

##### Computer-wide and user policy settings

> Admin/root users can overwrite computer-wide and user policy settings using `-settingsfile`,

Author: ```suggestion
Definitions:
  - System Lock-down mode:  When WDAC or AppLocker force PowerShell into Constrained Language mode and only trusted code runs in Full Language mode.  See [PowerShell Constrained Language Mode](https://devblogs.microsoft.com/powershell/powershell-constrained-language-mode/)
  
Admin/root users can overwrite computer-wide and user policy settings using `-settingsfile`,
```

> only when not in System Lock-down mode.

Reviewer 2: What is System Lock-down mode? Can we insert a comment or a doc reference?

| File -> Application-Startup | pwsh -settingsfile `somepath/powershell.config.json` | pwsh -settingsfile `somepath/powershell.config.json` |
| File -> Application-Wide    | $apphome\powershell.config.json                      | $apphome/powershell.config.json                      |
| File -> User-Wide           | %APPDATA%\powershell.config.json                     | ~/powershell.config.json                             |
| File -> Computer-Wide       | %ProgramFiles%\PowerShell\powershell.config.json     | /opt/Microsoft/powershell/powershell.config.json     |
| GPO -> User Config          | HKCU\Software\PowerShellCore                         | See `Comment A` below                                |
| GPO -> Computer Config      | HKLM\Software\PowerShellCore                         | /etc/powershell.config.json                          |

### Configuration settings

A set of configuration settings in GPO scheme and file scheme for policy settings and regular settings is the same. This allows to discover and configure settings in the simplest and fastest way.

> #### Registry keys and settings

Reviewer 2: Make sense to mark options which need to fallback to Windows PowerShell?

Author: All policies would allow this.  Your document didn't document policies, so I didn't include them yet.

Author: I have a few question about your comments.  I'll leave this open and update this after those replies.

        "EnableScriptBlockLogging": false
      },
      "ScriptExecution": {
        "ExecutionPolicy": "RemoteSigned",
        "PipelineMaxStackSizeMB": 10
      },
      "Transcription": {
        "EnableTranscripting": true,
        "EnableInvocationHeader": true,
        "OutputDirectory": "c:\\tmp"
      },
      "UpdatableHelp": {
        "DefaultSourcePath": "f:\\temp"
      }
    },

    "PoliciesSettings": {
      ...
    }
  },

  "OtherPowerShellApplication": {
    "RegularSettings": {
      ...
    },
    "PolicySettings": {
      ...
    }
}
```

## Alternate Proposals and Considerations

### Automatically resolve Windows PowerShell policy conflicts

We could attempt to resolve policy conflicts between PowerShell 7 policy and Windows PowerShell policy.
This would make the `Precedence for Policy settings` not just a simple list but a complex set of rules that would not be easily understood.  See [this conversation](https://github.com/PowerShell/PowerShell/issues/9309?#issuecomment-480643922).

### Allowing environment variable in the JSON

A new RFC should be drafted about how to allow environment variables in the JSON.

> This would allow consistent files across platforms.

Reviewer 2: Could you clarify do we mean env variables for the config file names or path names? If first this could be in the RFC.

## Specification

`PowerShell 7` should be configured using the following schemes:

- On Windows - Group Policy Objects (GPO), Group Policy Preferences (GPP) and settings files.
- On Unix - settings files.

The settings files have `JSON` format.

**Warning** The settings files differ from `PowerShell 7` _profile_ files, which are PowerShell scripts run at startup.

Configuration schemes allow to customize `PowerShell 7` in the most flexible way:

- Enterprise system administrators can use GPO,
    GPP and computer-wide settings files to apply approved configuration settings and mandatory security settings in a centralized manner.
    The same settings can be applied at user, application or startup levels.
- Developers and consumers can use user, application and startup level settings files.

### Configuration defaults

PowerShell 7 has hard-coded defaults for all configuration options.

The default values must be `secure-by-default`.

For release versions hard-coded defaults must be the same as ones in re-installed configuration files. For preview versions they may vary (ex., enable experimental features and so on).

> If during startup PowerShell 7 cannot read system configuration files it fails to startup.

Author: ```suggestion
System configuration includes security sensitive setting,
and failing to read those setting could result in an insecure system.
So, if during startup, PowerShell 7 cannot read system configuration files,
it fails to startup.
```

          "Joe"
        ]
      },
      "ScriptBlockLogging": {
        "EnableScriptBlockInvocationLogging": true,
        "EnableScriptBlockLogging": false
      },
      "ScriptExecution": {
        "ExecutionPolicy": "RemoteSigned",
        "PipelineMaxStackSizeMB": 10
      },
      "Transcription": {
        "EnableTranscripting": true,
        "EnableInvocationHeader": true,
        "OutputDirectory": "c:\\tmp"
      },
      "UpdatableHelp": {
        "DefaultSourcePath": "f:\\temp"
      }
    },

    "PoliciesSettings": {
      ...
    }
  },

  "OtherPowerShellApplication": {
    "RegularSettings": {
      ...
    },
    "PolicySettings": {
      ...
    }
}
```

## Alternate Proposals and Considerations

### Automatically resolve Windows PowerShell policy conflicts

We could attempt to resolve policy conflicts between PowerShell 7 policy and Windows PowerShell policy.
This would make the `Precedence for Policy settings` not just a simple list but a complex set of rules that would not be easily understood.  See [this conversation](https://github.com/PowerShell/PowerShell/issues/9309?#issuecomment-480643922).

### Allowing environment variable in the JSON

> A new RFC should be drafted about how to allow environment variables in the JSON.

Author: ```suggestion
A new RFC should be drafted about how to allow environment variables in the values in the JSON.
```

Regular settings can be treated as default and recommended values.
Policy settings is high priority and overlap regular settings.
Policy settings are used by administrators to centrally manage applications.

| Location     | Policy settings                                           | Regular settings                                           |
|--------------|-----------------------------------------------------------|------------------------------------------------------------|
| File section | "PowerShell": { "PolicySettings": {...} }                 | "PowerShell": { "RegularSettings": {...} }                 |
| File section | "OtherPowerShellApplication": { "PolicySettings": {...} } | "OtherPowerShellApplication": { "RegularSettings": {...} } |
| Registry key | Software\Policies\PowerShellCore                          | Software\PowerShellCore                                    |

### Policy settings Setting Fall-Back

#### Motivation - Policy Setting Fall-Back

This is to allow Fall-back to Windows PowerShell policies.

#### Implementation

For Policy Settings,
each policy should have a `Use Windows PowerShell Policy` which will indicate that the policy should the read from

> `SOFTWARE\Policies\Microsoft\Windows\PowerShell` instead of `Software\Policies\PowerShellCore`.

Author: ```suggestion
`SOFTWARE\Policies\Microsoft\Windows\PowerShell` instead of `Software\Policies\PowerShellCore`.
The default in Group Policy is to have on policy, so it would not fall back to Windows PowerShell Policy, or apply PowerShell 7 policy.
```

| Registry key | Software\Policies\PowerShellCore                          | Software\PowerShellCore                                    |

### Policy settings Setting Fall-Back

#### Motivation - Policy Setting Fall-Back

This is to allow Fall-back to Windows PowerShell policies.

#### Implementation

For Policy Settings,
each policy should have a `Use Windows PowerShell Policy` which will indicate that the policy should the read from
`SOFTWARE\Policies\Microsoft\Windows\PowerShell` instead of `Software\Policies\PowerShellCore`.
The default in Group Policy is to have on policy, so it would not fall back to Windows PowerShell Policy, or apply PowerShell 7 policy.

### Precedence of applying settings

Because a configuration setting can be in several schemes, the setting wins according to the priority of its scheme.

#### Precedence for Policy settings in descending order

| Scheme                      | Windows                                              | Unix                                                 |
|-----------------------------|------------------------------------------------------|------------------------------------------------------|

> | GPO -> Computer Policy      | HKLM\Software\Policies\PowerShellCore                | /etc/powershell.config.json                          |

Author: ```suggestion
| GPO -> Computer Policy      | HKLM\Software\Policies\PowerShellCore                | /etc/powershell/powershell.config.json                         |
```

| GPO -> User Policy          | HKCU\Software\Policies\PowerShellCore                | See `Comment A` below                                |
| File -> Computer-Wide       | %ProgramFiles%/PowerShell/powershell.config.json     | /etc/powershell/powershell.config.json     |
| File -> Application-Startup | pwsh -settingsfile `somepath/powershell.config.json` | pwsh -settingsfile `somepath/powershell.config.json` |
| File -> User-Wide           | %APPDATA%/powershell.config.json                     | %XDG_CONFIG_HOME%/powershell.config.json             |

> | File -> Application-Wide    | $apphome/powershell.config.json                      | $apphome/powershell.config.json                      |

Author: ```suggestion
| File -> Application-Wide    | $PSHome/powershell.config.json                      | $PSHome/powershell.config.json                      |
```

|-----------------------------|------------------------------------------------------|------------------------------------------------------|
| GPO -> Computer Policy      | HKLM\Software\Policies\PowerShellCore                | /etc/powershell/powershell.config.json                         |
| GPO -> User Policy          | HKCU\Software\Policies\PowerShellCore                | See `Comment A` below                                |
| File -> Computer-Wide       | %ProgramFiles%/PowerShell/powershell.config.json     | /etc/powershell/powershell.config.json     |
| File -> Application-Startup | pwsh -settingsfile `somepath/powershell.config.json` | pwsh -settingsfile `somepath/powershell.config.json` |
| File -> User-Wide           | %APPDATA%/powershell.config.json                     | %XDG_CONFIG_HOME%/powershell.config.json             |
| File -> Application-Wide    | $apphome/powershell.config.json                      | $apphome/powershell.config.json                      |

Defaults:

`%APPDATA%` - `C:\Users\useraccount\AppData\Roaming`

`%XDG_CONFIG_HOME%` - `HOME/.config`

> `$PSHome` definition from [about_powershell_config](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_powershell_config?view=powershell-6) - The $PSHOME location is defined as the same directory as the executing System.Management.Automation.dll assembly.

Author: ```suggestion

`$PSHome` definition from [about_powershell_config](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_powershell_config?view=powershell-6) - The $PSHOME location is defined as the same directory as the executing System.Management.Automation.dll assembly.
```

---
RFC:  RFCnnnn
Author:  travisez13
Status:  Draft
SupercededBy:  N/A
Version: 0.1
Area: Engine
Comments Due: 6/30/2019
---

# `PowerShell Core` Policy

## Motivation

> Consumers, developers, and enterprise system administrators should be able to flexibly and reliable way to configure PowerShell 7.

Author: ```suggestion
Consumers, developers, and enterprise system administrators should be able to flexibly and reliable configure PowerShell 7.
```

## Acknowledgement

I based this off of @iSazonov 's RFC, for just a slightly different purpose.
[PR #111](https://github.com/PowerShell/PowerShell-RFC/pull/111)

## Specification

`PowerShell 7` should be configured using the following schemes:

- On Windows - Group Policy Objects (GPO), Group Policy Preferences (GPP) and settings files.
- On Unix - settings files.

The settings files have `JSON` format.

**Warning** The settings files differ from `PowerShell 7` _profile_ files, which are PowerShell scripts run at startup.

Configuration schemes allow to customize `PowerShell 7` in the most flexible way:

- Enterprise system administrators can use GPO,
    GPP and computer-wide settings files to apply approved configuration settings and mandatory security settings in a centralized manner.
    The same settings can be applied at user, application or startup levels.

> - Developers and consumers can use user, application and startup level settings files.

Author: See `Precedence for Policy settings in descending order`

### Configuration defaults

PowerShell 7 has hard-coded defaults for all configuration options.

The default values must be `secure-by-default`.

> For release versions hard-coded defaults must be the same as ones in re-installed configuration files. For preview versions they may vary (ex., enable experimental features and so on).

Author: ```suggestion
For release versions hard-coded defaults must be the same as ones in pre-installed configuration files. For preview versions they may vary (ex., enable experimental features and so on).
```

System configuration includes security sensitive setting,
and failing to read those setting could result in an insecure system.

> So, if during startup, PowerShell 7 cannot read system configuration files,

Author: ```suggestion
So, if during startup, PowerShell 7 cannot read files read from the Computer-Wide scope,
```

it fails to startup.

If during startup PowerShell 7 cannot read user configuration files it uses _hardcoded_ defaults.

If during operation PowerShell 7 cannot read configuration files it continue to use _current_ (runtime) configuration values.

### Settings locations

`PowerShell 7` settings are grouped into `Policy settings` and `Regular settings`.
Regular settings are normal configuration settings.
Regular settings can be treated as default and recommended values.

> Policy settings is high priority and overlap regular settings.

Author: ```suggestion
Policy settings is high precedence.
See [Precedence for Policy settings in descending order](#precedence-for-policy-settings-in-descending-order).
```

> Policy settings are used by administrators to centrally manage applications.

Reviewer 2: Does this mean that we will have policies for PowerShell hosted applications similar to those in setting file? In the case we need to reword to be more clear.

Author: Let me think about this.

Reviewer 2: The feature is great. This already is below. My request is just for the sentence.

Author: I just need to think about the wording.

| Location     | Policy settings                                           | Regular settings                                           |
|--------------|-----------------------------------------------------------|------------------------------------------------------------|
| File section | "PowerShell": { "PolicySettings": {...} }                 | "PowerShell": { "RegularSettings": {...} }                 |

> | File section | "OtherPowerShellApplication": { "PolicySettings": {...} } | "OtherPowerShellApplication": { "RegularSettings": {...} } |

Author: ```suggestion
```

| Registry key | Software\Policies\PowerShellCore                          | Software\PowerShellCore                                    |

### Policy settings Setting Fall-Back

#### Motivation - Policy Setting Fall-Back

> This is to allow Fall-back to Windows PowerShell policies.

Reviewer 2: ```suggestion
This is to allow fall-back to Windows PowerShell policies.
```

#### Implementation

For Policy Settings,
each policy should have a `Use Windows PowerShell Policy` which will indicate that the policy should the read from
`SOFTWARE\Policies\Microsoft\Windows\PowerShell` instead of `Software\Policies\PowerShellCore`.

> The default in Group Policy is to have on policy, so it would not fall back to Windows PowerShell Policy, or apply PowerShell 7 policy.

Reviewer 2: ```suggestion
The default in Group Policy is to have no policy, so it would not fall back to Windows PowerShell Policy, or apply PowerShell 7 policy.
```

> The default in Group Policy is to have on policy, so it would not fall back to Windows PowerShell Policy, or apply PowerShell 7 policy.

Author: ```suggestion
The default in Group Policy is to have no policy, so it would not fall back to Windows PowerShell Policy, or apply PowerShell 7 policy.
```

### Precedence of applying settings

Because a configuration setting can be in several schemes, the setting wins according to the priority of its scheme.

#### Precedence for Policy settings in descending order

| Scheme                      | Windows                                              | Unix                                                 |
|-----------------------------|------------------------------------------------------|------------------------------------------------------|
| GPO -> Computer Policy      | HKLM\Software\Policies\PowerShellCore                | /etc/powershell/powershell.config.json                         |

> | GPO -> User Policy          | HKCU\Software\Policies\PowerShellCore                | See `Comment A` below                                |

Author: ```suggestion
| GPO -> User Policy          | HKCU\Software\Policies\PowerShellCore                | See [`Comment A`](#comment-a) below                                |
```

| File -> Application-Startup | pwsh -settingsfile `somepath/powershell.config.json` | pwsh -settingsfile `somepath/powershell.config.json` |
| File -> User-Wide           | %APPDATA%/powershell.config.json                     | %XDG_CONFIG_HOME%/powershell.config.json             |
| File -> Application-Wide    | $apphome/powershell.config.json                      | $apphome/powershell.config.json                      |

Defaults:

`%APPDATA%` - `C:\Users\useraccount\AppData\Roaming`

`%XDG_CONFIG_HOME%` - `HOME/.config`

`$PSHome` definition from [about_powershell_config](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_powershell_config?view=powershell-6) - The $PSHOME location is defined as the same directory as the executing System.Management.Automation.dll assembly.
This applies to hosted PowerShell SDK instances as well.
#### Parameter `-settingsfile`

With `-settingsfile` parameter users can assign custom settings from the config file and overwrite user-wide and application-wide settings.

##### Computer-wide and user policy settings

Definitions:

>   - System Lock-down mode:  When WDAC or AppLocker force PowerShell into Constrained Language mode and only trusted code runs in Full Language mode.  See [PowerShell Constrained Language Mode](https://devblogs.microsoft.com/powershell/powershell-constrained-language-mode/)

Reviewer 2: Please use full name for WDAC.

Author: ```suggestion
  - System Lock-down mode:  When Windows Defender Application Control or AppLocker force PowerShell into Constrained Language mode and only trusted code runs in Full Language mode.  See [PowerShell Constrained Language Mode](https://devblogs.microsoft.com/powershell/powershell-constrained-language-mode/)
```

|                                  | ModuleLogging               | EnableModuleLogging                | DWORD  | Computer, Then User |
|                                  | ModuleLogging               | ModuleNames                        | String | Computer, Then User |
|                                  | ProtectedEventLogging       | EncryptionCertificate              | DWORD  | Computer Wide       |
|                                  | ScriptBlockLogging          | EnableScriptBlockInvocationLogging | DWORD  | Computer, Then User |
|                                  | ScriptBlockLogging          | EnableScriptBlockLogging           | DWORD  | Computer, Then User |
|                                  | Transcription               | EnableTranscripting                | DWORD  | Computer, Then User |
|                                  | Transcription               | EnableInvocationHeader             | DWORD  | Computer, Then User |
|                                  | Transcription               | OutputDirectory                    | String | Computer, Then User |

> |                                  | UpdatableHelp               | DefaultSourcePath                  | String | Computer Wide       |

Reviewer 3: This table is a bit confusing to me. Do `Software\Policies\PowerShellCore` and `Software\PowerShellCore` have the same sub keys?

Author: There are no current keys listed for policies.  I'll update the table as the current comment says.

I filed [#9632](https://github.com/PowerShell/PowerShell/issues/9632) on UpdatableHelp-DefaultSourcePath to make it allow User settings.

#### JSON file settings format

```json
{
  "PowerShell": {
    "RegularSettings": {
      "ConsoleSessionConfiguration": {
        "EnableConsoleSessionConfiguration": true,
        "ConsoleSessionConfigurationName": "name"
      },
      "ProtectedEventLogging": {
        "EnableProtectedEventLogging": false,
        "EncryptionCertificate": [
          "Joe"
        ]
      },
      "ScriptBlockLogging": {
        "EnableScriptBlockInvocationLogging": true,
        "EnableScriptBlockLogging": false
      },
      "ScriptExecution": {
        "ExecutionPolicy": "RemoteSigned",
        "PipelineMaxStackSizeMB": 10
      },
      "Transcription": {
        "EnableTranscripting": true,
        "EnableInvocationHeader": true,
        "OutputDirectory": "c:\\tmp"
      },
      "UpdatableHelp": {
        "DefaultSourcePath": "f:\\temp"
      }
    },

    "PoliciesSettings": {
      ...
    }
  },

>   "OtherPowerShellApplication": {

Author: Remove this section

    "RegularSettings": {
      ...
    },
    "PolicySettings": {
      ...
    }
}
```

## Alternate Proposals and Considerations

### Automatically resolve Windows PowerShell policy conflicts

> We could attempt to resolve policy conflicts between PowerShell 7 policy and Windows PowerShell policy.

Reviewer 3: Why would we need to resolve the policy conflicts at all? They are pointed by different registry keys.

Author: This is an argument that we shouldn't.

Author: The big part of this is actually to make the fall back to Windows PowerShell Policy explicit.  This is the alternative to this:  https://github.com/PowerShell/PowerShell-RFC/pull/180/files#diff-45c05001a45a6563a6c836a16cc9081cR71

> We could attempt to resolve policy conflicts between PowerShell 7 policy and Windows PowerShell policy.

Author: ```suggestion
#### Motivation

This is a description of the alternative to [Policy settings Setting Fall-Back](#policy-settings-setting-fall-back).
The main purpose of describing the alternative is to describe why it should not be pursued.

#### Description

We could attempt to resolve policy conflicts between PowerShell 7 policy and Windows PowerShell policy.
```

This would make the `Precedence for Policy settings` not just a simple list but a complex set of rules that would not be easily understood.  See [this conversation](https://github.com/PowerShell/PowerShell/issues/9309?#issuecomment-480643922).

### Allowing environment variable in the JSON

A new RFC should be drafted about how to allow environment variables in the values in the JSON.
This would allow consistent files across platforms.

### Comment A

> Mainly for Unix we'd add `Users` section to computer wide JSON file (`/etc/powershell.config.json`) to allow administrators set policies and regular settings on user level base

Reviewer 3: Curious about how this is done on Windows. `HKCU\Software\PowerShellCore` points to the registry hive for the current user, can an administrator set registry keys for a specific user, like Smith?

Author: They have to mount the hive.  The equivalent in Linux would be to put a file into the user's folder.  I took this from @iSazonov's RFC.

```json
{
  "PowerShell": {
    "RegularSettings": {
      ...
    },
    "PolicySettings": {
      ...
    },

>   "Users": {

Reviewer 3: Would it better to use separate configuration files for each user? Like `smith-powershell.config.json`

Author: I think we should have a separate RFC about adding this feature.  cc @iSazonov

---
RFC:  RFCnnnn
Author:  travisez13
Status:  Draft
SupercededBy:  N/A
Version: 0.1
Area: Engine
Comments Due: 6/30/2019
---

# `PowerShell Core` Policy

## Motivation

Consumers, developers, and enterprise system administrators should be able to flexibly and reliably configure PowerShell 7.

## Acknowledgement

I based this off of @iSazonov 's RFC, for just a slightly different purpose.
[PR #111](https://github.com/PowerShell/PowerShell-RFC/pull/111)

## Goals

1. Specify how PowerShell 7 will deal with having both Windows PowerShell and PowerShell Core Group Policy.
    - This is covered in [Policy settings Setting Fall-Back](#policy-settings-setting-fall-back).
1. Define how the `pwsh -settingsfile` switch should behave.

>     - This is covered in  [Parameter `-settingsfile`](#Parameter--settingsfile)

Author: What is the third?

>     - This is covered in  [Parameter `-settingsfile`](#Parameter--settingsfile)

Reviewer 2: ```
The main goal here is to describe various policy and setting fall-back behavior:
- When things like -settingsfile will be allowed to overwrite, what type of settings.
- When PowerShell Core will use Windows PowerShell Policy, and how.
- Where various policies/settings are read from.
```

## Definitions

- **Computer-Wide settings/policy**  - setting or policy applied to an operating system environment (OSE), affecting all users of the OSE.

> - **User settings/policy** - setting or policy applied to a specific user of an OSE, and not applied to the OSE as a whole.

Author: It would be `Per-User settings` . 
```suggestion
- **Per-User settings/policy** - setting or policy applied to a specific user of an OSE, and not applied to the OSE as a whole.
```

## Specification

`PowerShell 7` should be configured using the following schemes:

> - On Windows - Group Policy Objects (GPO), and settings files.

Author: Feel free to add another RFC to add GPP, but it was only adding confusion.

- On Unix - settings files.

The settings files have `JSON` format.

**Warning** The settings files differ from `PowerShell 7` _profile_ files, which are PowerShell scripts run at startup.

Configuration schemes allow to customize `PowerShell 7` in the most flexible way:

- Enterprise system administrators can use GPO,

>     GPP and computer-wide settings files to apply approved configuration settings and mandatory security settings in a centralized manner.

Author: ```suggestion
    Computer-wide settings files to apply approved configuration settings and mandatory security settings in a centralized manner.
```

    Most settings can be applied either to the user or computer-wide.
    We will cover the precedence of these in [Registry keys and settings](#registry-keys-and-settings).
- Developers and consumers can use user, or computer-wide level setting files.

### Configuration defaults

PowerShell 7 has hard-coded defaults for all configuration policies and settings.

The default values must be `secure-by-default`.

For release versions hard-coded defaults must be the same as ones in pre-installed configuration files. For preview versions they may vary (ex., enable experimental features and so on).

Computer-wide configuration includes security sensitive setting,
and failing to read those setting, for example if the file is locked or corrupted, could result in an insecure system.
So, if during startup, PowerShell 7 cannot read settings files from the Computer-Wide scope,

> it fails to startup.

Author: We cannot have secure by default when encryption is involved.   We cannot securely encrypt and have a way for people to decrypt.

> it fails to startup.

Reviewer 2: > it fails to startup

I agree with the behavior in lock-down systems but not by default. 
As I said above, this can cause unpredictable behavior - users will be at a loss as to why their application is not launched at random.

If during startup PowerShell 7 cannot read user configuration files it uses _hardcoded_ defaults.

PowerShell 7 does not update configuration values from modified configuration files during a given session after the configuration has been loaded.

### Settings locations

`PowerShell 7` settings are grouped into `Policy settings` and `Regular settings`.
Regular settings are normal configuration settings.
Regular settings can be treated as default and recommended values.
Policy settings have a higher precedence than regular settings.
See [Precedence for Policy settings in descending order](#precedence-for-policy-settings-in-descending-order).
Policy settings are used by administrators to centrally manage PowerShell.

| Location     | Policy settings                                           | Regular settings                                           |
|--------------|-----------------------------------------------------------|------------------------------------------------------------|
| File section | "PowerShell": { "PolicySettings": {...} }                 | "PowerShell": { "RegularSettings": {...} }                 |

> | Registry key | Software\Policies\PowerShellCore                          | Not Applicable                                             |

Author: This is out of scope, it only adds confusion for this RFC as the behavior of this would not be described.

### Policy settings Setting Fall-Back

#### Motivation - Policy Setting Fall-Back

> This is to allow fall-back to Windows PowerShell policies.

Author: ```suggestion
Help to transition from Windows PowerShell to PowerShell 7.
```

| GPO -> User Policy          | HKCU\Software\Policies\PowerShellCore                | %XDG_CONFIG_HOME%/powershell.config.json             |
| File -> Application-Startup | pwsh -settingsfile `somepath/powershell.config.json` | pwsh -settingsfile `somepath/powershell.config.json` |
| File -> Computer-Wide       | See [Moving configuration out of PSHome][moving]     | [Moving configuration out of PSHome][moving]         |
| File -> User-Wide           | %APPDATA%/powershell.config.json                     | %XDG_CONFIG_HOME%/powershell.config.json             |

Defaults:

`%APPDATA%` - `C:\Users\useraccount\AppData\Roaming`

`%XDG_CONFIG_HOME%` - `HOME/.config`

#### Parameter `-settingsfile`

With `-settingsfile` parameter users can assign custom settings from the config file and overwrite user-wide settings.

##### More definitions

- System Lock-down mode:
  When Windows Defender Application Control or AppLocker force PowerShell into Constrained Language mode and
  only trusted code runs in Full Language mode.
  See [PowerShell Constrained Language Mode](https://devblogs.microsoft.com/powershell/powershell-constrained-language-mode/)

> ##### Computer-wide and user policy settings

Author: ```suggestion
##### Computer-wide and per-user policy settings
```

> Admin/root users can overwrite computer-wide and user policy settings using `-settingsfile`,

Author: ```suggestion
Admin/root users can overwrite computer-wide and per-user policy settings using `-settingsfile`,
```

only when not in System Lock-down mode.

This will have performance impact on startup, but only when `-settingsfile` is specified.

> #### Precedence for User settings in descending order

Author: ```suggestion
#### Precedence for Per-user settings in descending order
```

| ConsoleSessionConfiguration | EnableConsoleSessionConfiguration  | DWORD  | User, then Computer |
| ConsoleSessionConfiguration | ConsoleSessionConfigurationName    | String | User, then Computer |
| ModuleLogging               | EnableModuleLogging                | DWORD  | Computer, Then User |
| ModuleLogging               | ModuleNames                        | String | Computer, Then User |
| ProtectedEventLogging       | EncryptionCertificate              | DWORD  | Computer Wide       |
| ScriptBlockLogging          | EnableScriptBlockInvocationLogging | DWORD  | Computer, Then User |
| ScriptBlockLogging          | EnableScriptBlockLogging           | DWORD  | Computer, Then User |
| Transcription               | EnableTranscripting                | DWORD  | Computer, Then User |
| Transcription               | EnableInvocationHeader             | DWORD  | Computer, Then User |
| Transcription               | OutputDirectory                    | String | Computer, Then User |
| UpdatableHelp               | DefaultSourcePath                  | String | Computer Wide       |

> I filed [#9632](https://github.com/PowerShell/PowerShell/issues/9632) on UpdatableHelp-DefaultSourcePath to make it allow User settings.

Author: ```suggestion
I filed [#9632](https://github.com/PowerShell/PowerShell/issues/9632) on UpdatableHelp-DefaultSourcePath to make it allow Per-user settings.
```

        "DefaultSourcePath": "f:\\temp"
      }
    },

    "PoliciesSettings": {
      ...
    }
  }
}
```

## Alternate Proposals and Considerations

### Automatically resolve Windows PowerShell policy conflicts

#### Motivation -  Automatically policy

This is a description of the alternative to [Policy settings Setting Fall-Back](#policy-settings-setting-fall-back).
The main purpose of describing the alternative is to describe why it should not be pursued.

#### Description

PowerShell could attempt to resolve policy conflicts between PowerShell 7 policy and Windows PowerShell policy.
This would make the `Precedence for Policy settings` not just a simple list but a complex set of rules that would not be easily understood.  See [this conversation](https://github.com/PowerShell/PowerShell/issues/9309?#issuecomment-480643922).

### Allowing environment variable in the JSON

A new RFC should be drafted about how to allow environment variables in the values in the JSON.
This would allow consistent files across platforms.

### Moving configuration out of PSHome

Per issues [9278](https://github.com/PowerShell/PowerShell/issues/9278) we need to move configuration out of PSHome,

> follow that issue for issues related to new locations of files.

Author: Can we use https://github.com/PowerShell/PowerShell/issues/9278 for this topic.

    Most settings can be applied either to the user or computer-wide.
    We will cover the precedence of these in [Registry keys and settings](#registry-keys-and-settings).
- Developers and consumers can use user, or computer-wide level setting files.

### Configuration defaults

PowerShell 7 has hard-coded defaults for all configuration policies and settings.

The default values must be `secure-by-default`.

For release versions hard-coded defaults must be the same as ones in pre-installed configuration files. For preview versions they may vary (ex., enable experimental features and so on).

Computer-wide configuration includes security sensitive setting,
and failing to read those setting, for example if the file is locked or corrupted, could result in an insecure system.
So, if during startup, PowerShell 7 cannot read settings files from the Computer-Wide scope,
it fails to startup.
Because registry reads are more atomic,
this is not an issue for group policy settings,
but if we faced the same issues for these settings the solution would be the similar.

> If during startup PowerShell 7 cannot read user configuration files it uses _hardcoded_ defaults.

Reviewer 4: We should emit a warning as the hardcoded defaults may be less restrictive

Author: This should also be logged.

Author: ```suggestion
If during startup PowerShell 7 cannot read user configuration files it uses _hardcoded_ defaults and emits a warning.
```

PowerShell 7 does not update configuration values from modified configuration files during a given session after the configuration has been loaded.

### Settings locations

> `PowerShell 7` settings are grouped into `Policy settings` and `Regular settings`.

Reviewer 4: Would `Regular Settings` be better called `Preference Settings`?

Author: What do you mean my Preferences here?  Preferences has a specific Group Policy meaning.  I would suggest avoiding this terminology as it is confusing.

Author: I'm thinking `non-policy settings` is a better term

#### Precedence for UpdatableHelp in descending order

Note, this is listed as `Computer` in [Registry keys and settings](#registry-keys-and-settings).

| Scheme                      | Windows                                              | Unix                                                 |
|-----------------------------|------------------------------------------------------|------------------------------------------------------|
| GPO -> Computer Config      | HKLM\Software\PowerShellCore                         |  See [Moving configuration out of PSHome][moving]                          |
| File -> Application-Startup | pwsh -settingsfile `somepath/powershell.config.json` | pwsh -settingsfile `somepath/powershell.config.json` |
| File -> Computer-Wide       | See [Moving configuration out of PSHome][moving]     | /opt/Microsoft/powershell/powershell.config.json     |

### Configuration settings

A set of configuration settings in GPO scheme and file scheme for policy settings and regular settings is the same. This allows to discover and configure settings in the simplest and fastest way.

#### Registry keys and settings

> Notes:

Reviewer 5: i'm wondering if it is possible to alter the setting in memory after it has been created.

### Settings locations

`PowerShell 7` settings are grouped into `Policy settings` and `Non-policy settings`.
Regular settings are normal configuration settings.

> Regular settings can be treated as default and recommended values.

Reviewer 2: Should `Non-policy` be used everywhere?
